File: dustroute/routing.py
from __future__ import annotations
from dataclasses import dataclass
from heapq import heappush,heappop
from math import inf
from .model import *
from .wire import update_wire_shapes

@dataclass(frozen=True)
class RouteResult:
    path:tuple[Pos,...]
    cost:float

@dataclass(frozen=True)
class RouterConfig:
    max_nodes:int=30000
    horizontal_cost:float=1.
    stair_cost:float=2.5
    new_support_cost:float=1.5

class RouteNotFound(RuntimeError): pass

def heuristic(a,b):return abs(a.x-b.x)+abs(a.y-b.y)+abs(a.z-b.z)

def moves(p):
    out=[]
    for dx,dz in ((1,0),(-1,0),(0,1),(0,-1)):
        for dy in (0,1,-1):out.append(Pos(p.x+dx,p.y+dy,p.z+dz))
    return out

def _routeable(world,p,start,goal):
    b=world.get(p)
    if p in (start,goal):
        return b.kind in (BlockKind.AIR,BlockKind.REDSTONE_WIRE)
    # Existing dust belongs to some already-routed net and is therefore an
    # obstacle. This prevents accidental shorts during local rerouting.
    return b.kind is BlockKind.AIR

def astar_route(world,start,goal,config=RouterConfig()):
    if not _routeable(world,start,start,goal):raise RouteNotFound(f"bad start {start}")
    if not _routeable(world,goal,start,goal):raise RouteNotFound(f"bad goal {goal}")
    pq=[(heuristic(start,goal),0,start)];serial=0;g={start:0.};prev={};n=0
    while pq:
        _,_,cur=heappop(pq)
        if cur==goal:
            path=[cur]
            while cur in prev:cur=prev[cur];path.append(cur)
            return RouteResult(tuple(reversed(path)),g[goal])
        n+=1
        if n>config.max_nodes:break
        for q in moves(cur):
            if not _routeable(world,q,start,goal):continue
            sp=q.offset(dy=-1);sb=world.get(sp)
            if sb.kind not in (BlockKind.AIR,BlockKind.SOLID,BlockKind.TRANSPARENT,BlockKind.REDSTONE_BLOCK):continue
            c=config.stair_cost if q.y!=cur.y else config.horizontal_cost
            if sb.kind is BlockKind.AIR:c+=config.new_support_cost
            ng=g[cur]+c
            if ng>=g.get(q,inf):continue
            g[q]=ng;prev[q]=cur;serial+=1;heappush(pq,(ng+heuristic(q,goal),serial,q))
    raise RouteNotFound(f"No route {start}->{goal}")

def _facing(a,b):
    if a.y!=b.y:return None
    d=(b.x-a.x,b.z-a.z)
    return {(1,0):Facing.EAST,(-1,0):Facing.WEST,(0,1):Facing.SOUTH,(0,-1):Facing.NORTH}.get(d)

def materialize_route(world,result,support_kind=BlockKind.SOLID):
    for p in result.path:
        sp=p.offset(dy=-1)
        if world.get(sp).kind is BlockKind.AIR:world.set(sp,Block(support_kind))
        if world.get(p).kind is BlockKind.AIR:world.place(BlockKind.REDSTONE_WIRE,p.x,p.y,p.z)
    update_wire_shapes(world)

def insert_repeaters(world,path,max_wire_run=14,delay=1):
    reps=[];run=0
    for i in range(1,len(path)-1):
        run+=1
        if run<max_wire_run:continue
        fi=_facing(path[i-1],path[i]);fo=_facing(path[i],path[i+1])
        if fi is None or fi!=fo:continue
        p=path[i]
        world.place(BlockKind.REPEATER,p.x,p.y,p.z,facing=fo,delay=delay)
        reps.append(p);run=0
    update_wire_shapes(world);return tuple(reps)

def route_place_and_refresh(world,start,goal,config=RouterConfig(),max_wire_run=14):
    r=astar_route(world,start,goal,config)
    materialize_route(world,r)
    reps=insert_repeaters(world,r.path,max_wire_run)
    return r,reps

File: crates/dustroute-translate/src/routing.rs
use std::cmp::Reverse;
use std::collections::{BTreeMap, BinaryHeap};
use std::error::Error;
use std::fmt::{Display, Formatter};

use crate::world::{Block, BlockKind, Facing, Pos, World};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RouteResult {
    pub path: Vec<Pos>,
    /// Cost in tenths; integer costs keep queue ordering deterministic.
    pub cost: u32,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct RouterConfig {
    pub max_nodes: usize,
    pub horizontal_cost: u32,
    pub stair_cost: u32,
    pub new_support_cost: u32,
}

impl Default for RouterConfig {
    fn default() -> Self {
        Self {
            max_nodes: 30_000,
            horizontal_cost: 10,
            stair_cost: 25,
            new_support_cost: 15,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum RouteNotFound {
    InvalidEndpoint(Pos),
    Exhausted {
        start: Pos,
        goal: Pos,
        expanded: usize,
    },
}

impl Display for RouteNotFound {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::InvalidEndpoint(pos) => {
                write!(f, "invalid route endpoint {},{},{}", pos.x, pos.y, pos.z)
            }
            Self::Exhausted {
                start,
                goal,
                expanded,
            } => write!(f, "no route {start:?} -> {goal:?} after {expanded} nodes"),
        }
    }
}

impl Error for RouteNotFound {}

fn heuristic(a: Pos, b: Pos) -> u32 {
    a.x.abs_diff(b.x) + a.y.abs_diff(b.y) + a.z.abs_diff(b.z)
}

fn moves(pos: Pos) -> impl Iterator<Item = Pos> {
    [(1, 0), (-1, 0), (0, 1), (0, -1)]
        .into_iter()
        .flat_map(move |(dx, dz)| [-1, 0, 1].into_iter().map(move |dy| pos.offset(dx, dy, dz)))
}

fn routeable(world: &World, pos: Pos, start: Pos, goal: Pos) -> bool {
    let kind = world.kind_at(pos);
    if pos == start || pos == goal {
        matches!(kind, BlockKind::Air | BlockKind::RedstoneWire)
    } else {
        kind == BlockKind::Air
    }
}

pub fn astar_route(
    world: &World,
    start: Pos,
    goal: Pos,
    config: RouterConfig,
) -> Result<RouteResult, RouteNotFound> {
    if !routeable(world, start, start, goal) {
        return Err(RouteNotFound::InvalidEndpoint(start));
    }
    if !routeable(world, goal, start, goal) {
        return Err(RouteNotFound::InvalidEndpoint(goal));
    }
    let mut queue = BinaryHeap::from([Reverse((heuristic(start, goal) * 10, 0_u64, start))]);
    let mut serial = 0_u64;
    let mut costs = BTreeMap::from([(start, 0_u32)]);
    let mut previous = BTreeMap::new();
    let mut expanded = 0;
    while let Some(Reverse((_, _, current))) = queue.pop() {
        if current == goal {
            let mut path = vec![current];
            let mut cursor = current;
            while let Some(parent) = previous.get(&cursor).copied() {
                cursor = parent;
                path.push(cursor);
            }
            path.reverse();
            return Ok(RouteResult {
                path,
                cost: costs[&goal],
            });
        }
        expanded += 1;
        if expanded > config.max_nodes {
            break;
        }
        for next in moves(current) {
            if !routeable(world, next, start, goal) {
                continue;
            }
            let support = world.kind_at(next.offset(0, -1, 0));
            if !matches!(
                support,
                BlockKind::Air
                    | BlockKind::Solid
                    | BlockKind::Transparent
                    | BlockKind::RedstoneBlock
            ) {
                continue;
            }
            let step = if next.y == current.y {
                config.horizontal_cost
            } else {
                config.stair_cost
            };
            let new_cost = costs[&current]
                + step
                + if support == BlockKind::Air {
                    config.new_support_cost
                } else {
                    0
                };
            if costs.get(&next).is_some_and(|known| new_cost >= *known) {
                continue;
            }
            costs.insert(next, new_cost);
            previous.insert(next, current);
            serial += 1;
            queue.push(Reverse((
                new_cost + heuristic(next, goal) * 10,
                serial,
                next,
            )));
        }
    }
    Err(RouteNotFound::Exhausted {
        start,
        goal,
        expanded,
    })
}

pub fn materialize_route(world: &mut World, route: &RouteResult, support_kind: BlockKind) {
    for pos in &route.path {
        let support = pos.offset(0, -1, 0);
        if world.kind_at(support) == BlockKind::Air {
            world.set(support, Block::new(support_kind));
        }
        if world.kind_at(*pos) == BlockKind::Air {
            world.place(BlockKind::RedstoneWire, *pos);
        }
    }
}

fn facing(a: Pos, b: Pos) -> Option<Facing> {
    if a.y != b.y {
        return None;
    }
    match (b.x - a.x, b.z - a.z) {
        (1, 0) => Some(Facing::East),
        (-1, 0) => Some(Facing::West),
        (0, 1) => Some(Facing::South),
        (0, -1) => Some(Facing::North),
        _ => None,
    }
}

pub fn insert_repeaters(
    world: &mut World,
    path: &[Pos],
    max_wire_run: usize,
    delay: u8,
) -> Vec<Pos> {
    let mut repeaters = Vec::new();
    let mut run = 0;
    for index in 1..path.len().saturating_sub(1) {
        run += 1;
        if run < max_wire_run {
            continue;
        }
        let incoming = facing(path[index - 1], path[index]);
        let outgoing = facing(path[index], path[index + 1]);
        if incoming.is_none() || incoming != outgoing {
            continue;
        }
        let block = world.place(BlockKind::Repeater, path[index]);
        block.facing = outgoing;
        block.delay = Some(delay);
        repeaters.push(path[index]);
        run = 0;
    }
    repeaters
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn routes_and_refreshes_long_line() {
        let mut world = World::new();
        world.fill(
            Pos::new(0, 0, 0),
            Pos::new(40, 0, 0),
            Block::new(BlockKind::Solid),
        );
        let result = astar_route(
            &world,
            Pos::new(1, 1, 0),
            Pos::new(39, 1, 0),
            RouterConfig::default(),
        )
        .unwrap();
        materialize_route(&mut world, &result, BlockKind::Solid);
        let repeaters = insert_repeaters(&mut world, &result.path, 14, 1);
        assert!(repeaters.len() >= 2);
        assert!(world.validate_supports().is_ok());
    }
}
